test/integration/http2_integration_test.cc

```diff
@@ -253,4 +255,193 @@ TEST_P(Http2IntegrationTest, SimultaneousRequestWithBufferLimits) {
+    std::function<void(IntegrationStreamDecoder&)> cb) {
+  TestRandomGenerator rand;
+  const uint32_t num_requests = 50;
+  std::vector<Http::StreamEncoder*> encoders;
+  std::vector<IntegrationStreamDecoderPtr> responses;
+  std::vector<FakeStreamPtr> upstream_requests;
+
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  for (uint32_t i = 0; i < num_requests; ++i) {
+    responses.push_back(IntegrationStreamDecoderPtr{new IntegrationStreamDecoder(*dispatcher_)});
+    encoders.push_back(&codec_client_->startRequest(headers, *responses[i]));
+    codec_client_->sendData(*encoders[i], request_bytes, true);
+  }
+
+  for (uint32_t i = 0; i < num_requests; ++i) {
+    auto fake_upstream_connection =
+        FakeUpstream::waitForHttpConnection(*dispatcher_, fake_upstreams_);
+    // As data and streams are interwoven, make sure waitForNewStream()
+    // ignores incoming data and waits for actual stream establishment.
+    upstream_requests.push_back(fake_upstream_connection->waitForNewStream(true));
+    upstream_requests.back()->setAddServedByHeader(true);
+    fake_upstream_connections_.push_back(std::move(fake_upstream_connection));
+  }
+
+  for (uint32_t i = 0; i < num_requests; ++i) {
+    upstream_requests[i]->waitForEndStream(*dispatcher_);
+    upstream_requests[i]->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+    upstream_requests[i]->encodeData(rand.random() % (1024 * 2), true);
+  }
+
+  for (uint32_t i = 0; i < num_requests; ++i) {
+    responses[i]->waitForEndStream();
+    EXPECT_TRUE(upstream_requests[i]->complete());
+    EXPECT_EQ(request_bytes, upstream_requests[i]->bodyLength());
+
+    EXPECT_TRUE(responses[i]->complete());
+    cb(*responses[i]);
+  }
+}
+
+TEST_P(Http2RingHashIntegrationTest, CookieRoutingNoCookieNoTtl) {
+  config_helper_.addConfigModifier([&](envoy::api::v2::filter::HttpConnectionManager& hcm) -> void {
+    auto* hash_policy = hcm.mutable_route_config()
+                            ->mutable_virtual_hosts(0)
+                            ->mutable_routes(0)
+                            ->mutable_route()
+                            ->add_hash_policy();
+    auto* cookie = hash_policy->mutable_cookie();
+    cookie->set_name("foo");
+  });
+
+  std::set<std::string> served_by;
+  sendMultipleRequests(
+      1024,
+      Http::TestHeaderMapImpl{{":method", "POST"},
+                              {":path", "/test/long/url"},
+                              {":scheme", "http"},
+                              {":authority", "host"}},
+      [&](IntegrationStreamDecoder& response) {
+        EXPECT_STREQ("200", response.headers().Status()->value().c_str());
+        EXPECT_TRUE(response.headers().get(Http::Headers::get().SetCookie) == nullptr);
+        served_by.insert(
+            response.headers().get(Http::LowerCaseString("x-served-by"))->value().c_str());
+      });
+  EXPECT_EQ(served_by.size(), num_upstreams_);
```

Comment:
I'm a bit twitchy about this as it's technically non-deterministic and eventually you'll get an unlucky hash.  Think it's worth making num_upstreams_ non-const and bringing it down to 2 for this test or unlikely enough as-is?